Samples/ObservableScreen/Sources/CounterWorkflow.swift

```diff
@@ -0,0 +1,102 @@
+import Foundation
+import Workflow
+import WorkflowSwiftUI
+import WorkflowUI
+
+struct CounterWorkflow: Workflow {
+    // Dependencies from parent.
+    let info: CounterInfo
+    let resetToken: ResetToken
+    let initialValue: Int
+    let maxValue: Int?
+
+    @ObservableState
+    struct State {
+        var count: Int
+        var maxValue: Int?
+
+        var info: CounterInfo
+
+        var boundedCount: Int {
+            if let maxValue {
+                min(count, maxValue)
+            } else {
+                count
+            }
+        }
+    }
+
+    enum Action: WorkflowAction {
+        typealias WorkflowType = CounterWorkflow
+
+        case increment
+        case decrement
+
+        func apply(toState state: inout CounterWorkflow.State) -> CounterWorkflow.Output? {
+            // the max value is applied in the view, not here, so it's possible to increment
+            // above the max
+            switch self {
+            case .increment:
+                state.count += state.info.stepSize
+            case .decrement:
+                state.count -= state.info.stepSize
+            }
+            return nil
+        }
+    }
+
+    typealias Output = Never
+
+    init(info: CounterInfo, resetToken: ResetToken, initialValue: Int = 0, maxValue: Int? = nil) {
+        self.info = info
+        self.resetToken = resetToken
+        self.initialValue = initialValue
+        self.maxValue = maxValue
+    }
+
+    func makeInitialState() -> State {
+        State(count: initialValue, maxValue: maxValue, info: info)
+    }
+
+    func workflowDidChange(from previousWorkflow: CounterWorkflow, state: inout State) {
```

Comment:
Yeah, I'll have to cover this in the adoption docs, because it's definitely one of the biggest gotchas.

Parent dependencies need to go on your State so that we can track when they change and avoid body invalidations when they don't change, and `workflowDidChange` is the hook for updating state between two renders.

If a dependency is `@ObservableState` you can assign it directly and the macro handles checking changes. If it's not, you've got to do a comparison yourself. Otherwise, the property will be considered mutated on every render, and invalidate the body of any view that access that property on every render.

I think this is potentially something we could iterate on to make easier in the future, but for now I couldn't come up with anything better.

Reply:
I think we should definitely iterate on this! Wonder if a custom macro applied to the workflow itself could deal with this.

Reply:
+1 to spending some time iterating on this. Feels like it's almost certain to cause issues.

Reply:
Open to suggestions on this, but I won't have any additional iterations in this PR. Perhaps after we see some real usage, common patterns will be revealed.

Reply:
I agree that a macro on the workflow type seems like a promising way reduce boilerplate here, but is a deep enough problem that it should be left for a future iteration.

This is a great example case for that problem, where the workflow's props are a mix of `ObservableState` and non-OS types.